src/raft.rs

```diff
@@ -2457,4 +2559,53 @@ impl<T: Storage> Raft<T> {
         to_send.set_entries(req.take_entries());
         Some(to_send)
     }
+
+    /// Reduce size of 'ents' from uncommitted size.
+    pub fn reduce_uncommitted_size(&mut self, ents: &[Entry]) {
+        // fast path for non-leader endpoint
+        if self.state != StateRole::Leader {
+            return;
+        }
+
+        if !self.uncommitted_state.maybe_reduce_uncommitted_size(ents) {
+            // this will make self.uncommitted size not accurate.
+            // but in most situation, this behaviour will not cause big problem
+            warn!(
+                self.r.logger,
+                "try to reduce uncommitted size less than 0, first index of pending ents is {}",
+                ents[0].get_index()
+            );
+        }
+    }
+
+    /// Increase size of 'ents' to uncommitted size. Return true when size limit
+    /// is satisfied. Otherwise return false and uncommitted size remains unchanged.
+    /// For raft with no limit(or non-leader raft), it always return true.
+    pub fn maybe_increase_uncommitted_size(&mut self, ents: &[Entry]) -> bool {
+        // fast path for non-leader endpoint
+        if self.state != StateRole::Leader {
```

Comment:
Because the method will only be called when it's leader, the fast path is actually slow path and causes wrong branch detection. I suggest to remove it or add a `debug_assert`.